src/debug/telemetry/telemetry.c

```diff
@@ -0,0 +1,145 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2024 Intel Corporation. All rights reserved.
+//
+// Author: Tobiasz Dryjanski <tobiaszx.dryjanski@intel.com>
+
+#include <zephyr/debug/sparse.h>
+
+#include <sof/audio/module_adapter/module/generic.h>
+#include <sof/debug/telemetry/telemetry.h>
+
+#include <adsp_debug_window.h>
+#include <errno.h>
+#include <limits.h>
+#include <mem_window.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+/* Systic variables, one set per core */
+static int telemetry_systick_counter[CONFIG_MAX_CORE_COUNT];
+static int telemetry_prev_ccount[CONFIG_MAX_CORE_COUNT];
+static int telemetry_perf_period_sum[CONFIG_MAX_CORE_COUNT];
+static int telemetry_perf_period_cnt[CONFIG_MAX_CORE_COUNT];
+static struct telemetry_perf_queue telemetry_perf_queue[CONFIG_MAX_CORE_COUNT];
+
+static void telemetry_perf_queue_append(struct telemetry_perf_queue *q, size_t element)
+{
+	if (!q->full) {
+		q->elements[q->index] = element;
+		q->sum += element;
+		q->index++;
+		q->size++;
+		if (q->index >= SOF_AVG_PERF_MEAS_DEPTH) {
+			q->index = 0;
+			q->size = SOF_AVG_PERF_MEAS_DEPTH;
+			q->full = true;
+		}
+	} else {
+		/* no space, pop tail */
+		q->sum -= q->elements[q->index];
+		/* replace tail */
+		q->elements[q->index] = element;
+		q->sum += element;
+		/* move tail */
+		q->index++;
+		if (q->index >= SOF_AVG_PERF_MEAS_DEPTH)
+			q->index = 0;
+	}
+}
+
+static size_t telemetry_perf_queue_avg(struct telemetry_perf_queue *q)
+{
+	if (!q->size)
+		return 0;
+	return q->sum / q->size;
+}
+
+int telemetry_init(void)
+{
+	/* systick_init */
+	uint8_t slot_num = SOF_DW_TELEMETRY_SLOT;
+	volatile struct adsp_debug_window *window = ADSP_DW;
+	struct telemetry_wnd_data *wnd_data = (struct telemetry_wnd_data *)ADSP_DW->slots[slot_num];
+	struct system_tick_info *systick_info =
+			(struct system_tick_info *)wnd_data->system_tick_info;
+
+	window->descs[slot_num].type = ADSP_DW_SLOT_TELEMETRY;
+	window->descs[slot_num].resource_id = 0;
+	wnd_data->separator_1 = 0x0000C0DE;
+
+	/* Zero values per core */
+	for (int i = 0; i < CONFIG_MAX_CORE_COUNT; i++) {
+		systick_info[i].count = 0;
+		systick_info[i].last_sys_tick_count = 0;
+		systick_info[i].max_sys_tick_count = 0;
+		systick_info[i].last_ccount = 0;
+		systick_info[i].avg_utilization = 0;
+		systick_info[i].peak_utilization = 0;
+		systick_info[i].peak_utilization_4k = 0;
+		systick_info[i].peak_utilization_8k = 0;
+	}
+	return 0;
+}
+
+void telemetry_update(uint32_t begin_ccount, uint32_t current_ccount)
+{
+	int prid = cpu_get_id();
+
+	++telemetry_systick_counter[prid];
+
+	struct telemetry_wnd_data *wnd_data =
+		(struct telemetry_wnd_data *)ADSP_DW->slots[SOF_DW_TELEMETRY_SLOT];
+	struct system_tick_info *systick_info =
+		(struct system_tick_info *)wnd_data->system_tick_info;
+
+	systick_info[prid].count = telemetry_systick_counter[prid];
+	systick_info[prid].last_sys_tick_count = current_ccount - begin_ccount;
```

Comment:
Just to whine a bit, the terminology here seems pretty mixed up:

* The code right here is calling it a "ccount", which implies that it's sourced from the local core's CCOUNT SR.  It isn't[1], it's derived from sof_cycle_get_64(), which is a wrapper around Zephyr k_cycle_get_64(), which is returning the HDA wall clock time.

* It's then being assigned to a variable called "tick_count", implying that it might be a Zephyr tick (12 kHz) or SOF scheduler quantum (1 kHz).  It's not those either.

* The PR is using the term "systick", which isn't a standard thing AFAIK?  Where it's used in the industry, it generally is a reference to the ARM Cortex M internal timer device.  It's not that either. :)

[1] And really can't be; CCOUNT isn't synchronized between cores and thus can't be used as a timestamp.

Reply:
Since this is a port from the reference firmware I just kept the original field names. 

About the ccount function, earlier in this review there was a talk about this. Originally I used  ` xthal_get_ccount();` but I was told to use a Zephyr side function, and it ended on using `sof_cycle_get_64()`. Is it really a good replacement? I'm guessing the scaling may be wrong.

About your other question about DSP clock rate: had to check the tests how it was handled and yes, you are right, clock rate was hard coded per platform. It seems that "the other side has to just know" was the idea when this was originally implemented.

Reply:
> Since this is a port from the reference firmware I just kept the original field names.

The reference firmware is not based on a separate project like Zephyr and it's not multi-vendors either. So it has totally different collision and consistency constraints.

https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#rule-a-3-macro-name-collisions

Preserving register names and other vendor-specific symbols is great. Preserving names of high level concepts like `tick_count` is not good. It's much more important for these to be consistent with Zephyr and industry standards than with some closed-sourced, self contained project.

So I think this is the key question: is this new feature SOF_TELEMETRY or rather SOF_INTEL_TELEMETRY? Very important question.

Reply:
@andyross Used ccount (xthal_get_ccount) before i was told to change it to sof_cycle_get_64. It was all made with ccount in mind so I'm not sure if it's a good replacement. What is the actual difference between output values of those two? 
 Also why is ccount not good for timestamps? Isn't there one ccount value shared for all cores?

Reply:
I think thing is still open. We already have overload of "tick" as this is widely used term with a very specific meaning in Zephyr, and on SOF side, IPC4 has added definition of tick in base_fw.h that is not the same as Zephyr. Now in this code, a third type of "tick" is added.

Given this is a very hardware specific interface to begin with, I'd actually just use "ccount" directly to make this explicit. CCOUNT is not shared by cores, and with DSP clock rate changes, clock/power gating, the ccount will not be a usable timestamp even for a single core. That's why we use sof_cycle_get_64() (which is not ccount) for timestamps in SOF. But if the telemetry interface expects ccount, then that's what you need to fill.

But what are the systick fields here? I've seen some uses of "DSP ticks" (= CCOUNT), so I guess that is possible. I'd then add comments to the code as this will be very confusing as this is the third type of "tick" in the same codebase.

If you compile Zephyr with CONFIG_TIMING_FUNCTIONS, then you can get CCOUNT in a more portable way with timing_counter_get() on xtensa. However, we disable CONFIG_TIMING_FUNCTIONS by default in the build and only use it performance overlay builds. so this will require some additional investigation to added (e.g. @abonislawski disabled the DSP clock gating for perf overlay -> this is not ok for telemetry if it needs to be enabled by default).



Use of word "tick" is very confusing as we have established meaning of ticks in Zephyr and before this PRGiven this is a very hardware specific interface

Reply:
@tobonex At a minimum, we need a comment to explain the tick semantics and why we are putting non-ccount values into variable called ccount. I can accept to do this on purpose, but you need to have a comment that explains why this is done.

Reply:
@kv2019i  One more thing here, if sof_cycle_get_64 is the final pick here, what is it's tick rate compared to ccount? I'll probably need to rescale the timestamp if it's different.
As for the names, I'll probably just rename the fields.

Reply:
@tobonex it will depend on build target, but with most SOF targets (including all Intel ones), sof_cycle_get_64 will be mapped to k_cycle_get_64() which is defined as "* This routine returns the current time in 64-bits, as measured by the                                                                                                   system's hardware clock, if available.  "
This again depends on the build and target, but on Intel platforms this is the SoC timer which runs at rate of CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC (typically 38.4Mhz).

But it really depends on how the telemetry interface is defined. It is defined to track number of DSP cycles, then you have to use ccount() (or similar).

If it's tracking time, then you have to use sof_cycle_get_64() and provide out-of-band means for telemetry use to discover the value of CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC. 

Neither is really correct or incorrect, and if the telemetry interface documents the semantics, both are ok to me. This is only a blocker if semantics say one thing and implementation does something else.

Reply:
@kv2019i Originally (in reference FW) it used ccount, and that was my initial idea here. Tests compare those measurements to the dsp clock rate.
Our tests basically compute the resulting percentage like this:
core_clock = 40000000
max_kpcs = core_clock/1000  
percentage=(ccount/max_kpcs)*100 
Where ccount is the one we measure here for the zephyr_ll_run, read by host from memwindow.

Other reviewers said to use sof_cycle_get_64, but I still think ccount is more appropriate.

Reply:
It's very dependent on what you are measuring. The DSP clock can be changed at runtime and clock can be gated. But that's not necessary a problem if you calculate percentage of how the load compares to maximum the DSP can do. If you want to calculate how much time DSP spent active, then this formula won't work, but that's not necessarily a problem.

Not sure who recommended sof_cycle_get_64, I can't see in the public comments. It really al depends on what the interface tries to measure. Both ccount and sof_cycle_get_64 will work, but you are reporting different things.

Reply:
I'll leave sof_cycle_get_64 here for now then. Systick will be revisited later anyway,  so there's time for potential changes.
About the naming, I changed variable names to be more generic. Added a comment at systick struct to differentiate this "tick" from other ticks.